checker/src/main/java/org/checkerframework/checker/optional/OptionalAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,122 @@
+import org.checkerframework.framework.flow.CFAbstractAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.TreeUtils;
+
+/** OptionalAnnotatedTypeFactory for the Optional Checker. */
+public class OptionalAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+  /** The element for java.util.Optional.map(). */
+  private final ExecutableElement optionalMap;
+
+  /** The @{@link Present} annotation. */
+  protected final AnnotationMirror PRESENT = AnnotationBuilder.fromClass(elements, Present.class);
+
+  /**
+   * Creates an OptionalAnnotatedTypeFactory.
+   *
+   * @param checker the Optional Checker associated with this type factory
+   */
+  public OptionalAnnotatedTypeFactory(BaseTypeChecker checker) {
+    super(checker);
+    postInit();
+    optionalMap = TreeUtils.getMethodOrNull("java.util.Optional", "map", 1, getProcessingEnv());
+  }
+
+  @Override
+  public AnnotatedTypeMirror getAnnotatedType(Tree tree) {
+    AnnotatedTypeMirror result = super.getAnnotatedType(tree);
+    optionalMapNonNull(tree, result);
+    return result;
+  }
+
+  /**
+   * If {@code tree} is a call to {@link java.util.Optional#map(Function)} whose argument is a
+   * method reference, then this method adds {@code @Present} to {@code type} if the following is
+   * true:
+   *
+   * <ul>
+   *   <li>The type of the receiver to {@link java.util.Optional#map(Function)} is {@code @Present},
+   *       and
+   *   <li>{@link #returnsNonNull(MemberReferenceTree)} returns true.
+   * </ul>
+   *
+   * @param tree a tree
+   * @param type the type of the tree, which may be side-effected by this method
+   */
+  private void optionalMapNonNull(Tree tree, AnnotatedTypeMirror type) {
+    if (!TreeUtils.isMethodInvocation(tree, optionalMap, processingEnv)) {
+      return;
+    }
+    MethodInvocationTree mapTree = (MethodInvocationTree) tree;
+    ExpressionTree argTree = mapTree.getArguments().get(0);
+    if (argTree.getKind() == Kind.MEMBER_REFERENCE) {
+      MemberReferenceTree memberReferenceTree = (MemberReferenceTree) argTree;
+      AnnotatedTypeMirror optType = getReceiverType(mapTree);
+      if (optType == null || !optType.hasEffectiveAnnotation(Present.class)) {
+        return;
+      }
+      if (returnsNonNull(memberReferenceTree)) {
+        type.replaceAnnotation(PRESENT);
+      }
+    }
+  }
+
+  /**
+   * Returns true if the return type of the function type of {@code memberReferenceTree} is
+   * non-null.
+   *
+   * @param memberReferenceTree a member reference
+   * @return true if the return type of the function type of {@code memberReferenceTree} is non-null
+   */
+  private boolean returnsNonNull(MemberReferenceTree memberReferenceTree) {
```

Comment:
I find `returnsNonNull` (and the documentation above) a bit misleading.  The given function can return null, but only if it is passed null.  Another way to say this is that the function never introduces a new null value.  Could you please adjust the method name and documentation accordingly?  Thank you!

Reply:
This method is only checking that the return type does not have an explicit nullable annotation.  If the return type is annotated with `@PolyNull`, then it still may return null even if the argument is non-null. Here's an example method signature with that property.
```
  @PolyNull Integer convertPoly(List<@PolyNull String> s);
```

Reply:
Thanks!  That renaming is helpful.

There is an inconsistency between the documentation (which says "true if `@Nullable`") and the name (which says "true if not `@Nullable`").  I suspect this is just a typo,  Could you make them consistent?  My preference would be to avoid negation in method names (& specifications), because negation makes reasoning more difficult.

Regarding the semantics:  is there a difference between "is not annotated as `@Nullable`" and "never introduces a *new* null value"?  I think they are the same, assuming that the code is fully annotated.  Noting this equivalence (or explicitly disclaiming it, if I am mistaken) would help to clarify the effect of the code.

Reply:
> Regarding the semantics: is there a difference between "is not annotated as @Nullable" and "never introduces a new null value"? I think they are the same, assuming that the code is fully annotated. Noting this equivalence (or explicitly disclaiming it, if I am mistaken) would help to clarify the effect of the code.

I don't know what "never introduces a new null value" means.  (What's a new null value? How does a method reference introduce a null value?) I think you mean that the function only returns null if its argument is also null.  That is not the same a method whose return type is not annotated with a nullable annotation.  Here's an example of method that is not annotated with a nullable annotation and returns null when its argument in non-null:

```
  @PolyNull Integer convertPoly(List<@PolyNull String> s) {
    for (String ins : s) {
      if (ins == null) {
        return null;
      }
    }
    return 0;
  }
```

Reply:
Thank you for this example!  It is helpful.  I indeed meant that the function only returns null if its argument is also null.  But, as you point out, `@PolyNull` might not be written on the formal parameter.  I had overlooked that.  That would need to be checked.

Could you add a note about this to the documentation?  Thanks.